controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,369 @@
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	routeList := &gatewayapiv1.HTTPRouteList{}
+	// Get all the routes having the gateway as parent
+	err = r.Client().List(ctx, routeList, client.MatchingFields{HTTPRouteParents: client.ObjectKeyFromObject(gw).String()})
+	logger.V(1).Info("gatewayAPITopologyFromGateway: list httproutes from gateway", "err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	rlpList := &kuadrantv1beta2.RateLimitPolicyList{}
+	// Get all the rate limit policies
+	// TODO(eastizle): Add index field??
+	err = r.Client().List(ctx, rlpList)
+	logger.V(1).Info("gatewayAPITopologyFromGateway: list rate limit policies", "err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	return common.NewKuadrantTopology(
+		[]*gatewayapiv1.Gateway{gw},
+		common.Map(routeList.Items, func(r gatewayapiv1.HTTPRoute) *gatewayapiv1.HTTPRoute { return &r }),
+		common.Map(rlpList.Items, func(p kuadrantv1beta2.RateLimitPolicy) common.KuadrantPolicy { return &p }),
+	), nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) WASMRateLimitPolicy(ctx context.Context, t *common.KuadrantTopology, rlp *kuadrantv1beta2.RateLimitPolicy, gw *gatewayapiv1.Gateway) (*wasm.RateLimitPolicy, error) {
+	gwHostnamesTmp := common.TargetHostnames(gw)
+	gwHostnames := common.Map(gwHostnamesTmp, func(str string) gatewayapiv1.Hostname { return gatewayapiv1.Hostname(str) })
+
+	route, err := r.RouteFromRLP(ctx, t, rlp, gw)
+	if err != nil {
+		return nil, err
+	}
+	if route == nil {
+		// no need to add the policy if there are no routes;
+		// a rlp can return no rules if all its limits fail to match any route rule
+		// or targeting a gateway with no "free" routes. "free" meaning no route with policies targeting it
+		return nil, nil
+	}
+
+	rules := rlptools.WasmRules(rlp, route)
+	if len(rules) == 0 {
+		// no need to add the policy if there are no rules; a rlp can return no rules if all its limits fail to match any route rule
+		return nil, nil
+	}
+
+	// narrow the list of hostnames specified in the route so we don't generate wasm rules that only apply to other gateways
+	// this is a no-op for the gateway rlp
+	hostnames := common.FilterValidSubdomains(gwHostnames, route.Spec.Hostnames)
+	if len(hostnames) == 0 { // it should only happen when the route specifies no hostnames
+		hostnames = gwHostnames
+	}
+
+	return &wasm.RateLimitPolicy{
+		Name:      client.ObjectKeyFromObject(rlp).String(),
+		Domain:    rlptools.LimitsNamespaceFromRLP(rlp),
+		Hostnames: common.HostnamesToStrings(hostnames), // we might be listing more hostnames than needed due to route selectors hostnames possibly being more restrictive
+		Service:   common.KuadrantRateLimitClusterName,
+		Rules:     rules,
+	}, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) RouteFromRLP(ctx context.Context, t *common.KuadrantTopology, rlp *kuadrantv1beta2.RateLimitPolicy, gw *gatewayapiv1.Gateway) (*gatewayapiv1.HTTPRoute, error) {
```

Comment:
Could this func belong to the `KuadrantTopology` type? I imagine that extracting the scope HTTPRoute for an AuthPolicy would look like pretty much the same, but I could be missing something.

Reply:
This method logic is simple: when the policy is targeting a (valid) route, return that route. If the policy is targeting a (valid) gateway, compute an "imaginary" route made up from all the "free" routes attached to the gateway. 

I do not think that this logic belongs to the kuadrant topology type which is only, and nothing else than a representation or view of the cluster state. The logic above is the implementation of the current kuadrant's design of policies targeting a gateway and its semantics. 

I agree, though, that this `RouteFromRLP` method can be generalized to accomodate any kuadrant policy as long as kuadrant implements the same approach for any policy targeting a gateway. 

Let me stick with this implementation for now and let's refactor and generalize when the authpolicy controller leverages the kuadrant policy here presented.

Reply:
I'd argue the function _already_ is generic enough for all kinds of Kuadrant policies. I think it's literally only the signature that makes it RLP-specific now, no?